### ISSUE TITLE
Log capturing in unit tests

### ISSUE THREAD

**Issue by dbrgn** [slog-rs/slog#256]
When writing unit tests, cargo captures output to stdout by default.

```rust
#[test]
fn print() {
    println!("Logging with println!");
}
```

This will not print the text to the terminal, unless `cargo test` is started with the `-- --nocapture` argument.

However, with slog this does not seem to work:

```rust
// main.rs

fn main() {
    println!("Hello, world!");
}

#[cfg(test)]
mod tests {
    use slog::{Drain, o, info};

    #[test]
    fn print() {
        println!("Logging with println!");
    }

    #[test]
    fn log_to_stdout() {
        let decorator = slog_term::TermDecorator::new().stdout().build();
        let drain = slog_term::CompactFormat::new(decorator).build();
        let log = slog::Logger::root(std::sync::Mutex::new(drain).fuse(), o!());
        info!(log, "Logging to stdout");
    }

    #[test]
    fn log_to_stderr() {
        let decorator = slog_term::TermDecorator::new().stderr().build();
        let drain = slog_term::CompactFormat::new(decorator).build();
        let log = slog::Logger::root(std::sync::Mutex::new(drain).fuse(), o!());
        info!(log, "Logging to stderr");
    }
}
```

Running this with `cargo test` will print:

```
$ cargo test -- --test-threads=1
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running target/debug/deps/asdf-5593e16ab0d57485

running 3 tests
test tests::log_to_stderr ... Feb 17 17:33:41.976 INFO Logging to stderr
ok
test tests::log_to_stdout ... Feb 17 17:33:41.977 INFO Logging to stdout
ok
test tests::print ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

Neither stdout nor stderr logs are captured.

I realize that this is probably due to some pecularity in how cargo handles capturing, but is there some kind of drain that logs in a cargo log capturing compatible way (possibly using `println`) that can be used in tests?

**Comment by Flakebi** [comment]
You need to use `let decorator = slog_term::PlainDecorator::new(std::io::stdout());`. The `TermDecorator` will print directly to the terminal if possible.

**Comment by dpc** [comment]
I might be wrong, but I think tests in Rust are not intercepting stdio, but only calls to `println` and alikes. I think someone had a slog Drain/serializer that is using `println` instead of writing to `stdio` directly to help with slog logging in tests.

**Comment by dpc** [comment]
Ha. It's in `slog-term` directly. https://github.com/slog-rs/term/blob/1f749568fe03f73be0142f2fdf9ac80ee961efc6/src/lib.rs#L1392

**Comment by dpc** [comment]
What's the point of writing all this documentation when no one reads it. :D https://docs.rs/slog-term/2.5.0/slog_term/struct.TestStdoutWriter.html

**Comment by Flakebi** [comment]
Right, I thought I got it working but I failed to run the test… Thanks!

**Comment by Flakebi** [comment]
Also, for future reference because I just fixed this, I needed to use `Mutex` instead of `Async` in `#[tokio::test]`:
```rust
let logger = {
	let decorator = slog_term::PlainDecorator::new(slog_term::TestStdoutWriter);
	let drain = Mutex::new(slog_term::FullFormat::new(decorator).build()).fuse();

	slog::Logger::root(drain, o!())
};
```

**Comment by dpc** [comment]
Async should work too, no? Particularly in `async` code I would recommend `async` drain, because you want to minimize the executor thread being stuck on io. Locking a mutex over a drain is OK in a slow synchronous code, but in async, I doubt it will be good enough.

**Comment by Flakebi** [comment]
It works, yes, but I get all the console output that I don’t want for succeeding unit tests. With the `Mutex` I only get the output when a test fails :)

**Comment by dpc** [comment]
Really? Why? Are you using both `slog_term::TestStdoutWriter` + `Async`?

**Comment by Flakebi** [comment]
Yes, the variant that works good for me is this:
```rust
#[tokio::test]
async fn test_things() -> Result<()> {
	let logger = {
		let decorator = slog_term::PlainDecorator::new(slog_term::TestStdoutWriter);
		let drain = Mutex::new(slog_term::FullFormat::new(decorator).build()).fuse();

		slog::Logger::root(drain, o!())
	};
	info!(logger, "Hi from async test");
	Ok(())
}
```

**Comment by Bathtor** [comment]
I know this has been closed for a while, but I'd like to save other people some headache and time, trying to figure out why this doesn't work with `Async` (in fact it doesn't work with `Mutex` either, if you are actually logging something from another thread):

https://github.com/rust-lang/rust/issues/42474

**TL;DR:** Rust's testing framework uses thread locals to mark that output should be captured, so creating any threads in a test case and passing a logger (or just even calling just `println!`) there is not going to work.

Perhaps this could be mentioned in the docs at some point?

**Comment by dpc** [comment]
Please fill a PR!

**Comment by Bathtor** [comment]
> Please fill a PR!

There you go: https://github.com/slog-rs/term/pull/32

**Comment by dpc** [comment]
Thanks!
